### PR TITLE
Simplifying lowering and abstract evaluation rules

### PR DESCRIPTION
**Pull request by dfm**
With relatively recent versions of JAX we can remove some of the boilerplate in the MLIR lowering rules.

### CODE REVIEW

File: src/jax_finufft/shapes.py [hunk 1]
```diff
@@ -116,10 +115,15 @@ def abstract_eval(source, *points, output_shape, **_):
...
     assert source.shape[0] == points[0].shape[0]
-    if output_shape is None:
+
+    if output_shape is None:  # Type 2
         assert source.ndim == 2 + ndim
-        return ShapedArray(source.shape[:2] + (points[0].shape[-1],), source_dtype)
-    else:
+        return source.update(
+            shape=source.shape[:2] + (points[0].shape[-1],), dtype=source_dtype
+        )
```
Comment (dfm) [inline src/jax_finufft/shapes.py:123]: This is an unrelated change that I noticed. We're no longer supposed to manually construct a `ShapedArray` here because it should return the same type as the input (tracer vs concrete value).

**Comment by dfm** [comment]
@lgarrison — Good point! It looks like the oldest version of JAX that this supports is [0.4.20](https://github.com/google/jax/releases/tag/jax-v0.4.20) from Nov 2023. This would make us somewhat bleeding edge, but I'm not too concerned about that. What do you think?

**Comment by lgarrison** [comment]
I think that sounds fine, we already have a release that works with older JAX on PyPI. LGTM!
